cli/util/progress_bar.rs

```diff
@@ -6,6 +6,299 @@ use indexmap::IndexSet;
+    progress_bars: Vec<ProgressBar>,
+  }
+
+  fn clear_previous_line() {
+    eprint!("\x1B[1A\x1B[0J");
+  }
+
+  impl ProgressBars {
+    /// Checks if progress bars are supported
+    pub fn are_supported() -> bool {
+      atty::is(atty::Stream::Stderr) && get_terminal_width().is_some()
+    }
+
+    /// Creates a new ProgressBars or returns None when not supported.
+    pub fn new() -> Option<Self> {
+      if ProgressBars::are_supported() {
+        Some(ProgressBars {
+          state: Arc::new(RwLock::new(InternalState {
+            drawer_id: 0,
+            progress_bar_counter: 0,
+            progress_bars: Vec::new(),
+          })),
+        })
+      } else {
+        None
+      }
+    }
+
+    pub fn add_progress(
+      &self,
+      message: String,
+      style: ProgressBarStyle,
+      total_size: usize,
+    ) -> ProgressBar {
+      let mut internal_state = self.state.write();
+      let id = internal_state.progress_bar_counter;
+      let pb = ProgressBar {
+        id,
+        progress_bars: self.clone(),
+        start_time: SystemTime::now(),
+        message,
+        size: total_size,
+        style,
+        pos: Arc::new(RwLock::new(0)),
+      };
+      internal_state.progress_bars.push(pb.clone());
+      internal_state.progress_bar_counter += 1;
+
+      if internal_state.progress_bars.len() == 1 {
+        self.start_draw_thread(&mut internal_state);
+      }
+
+      pb
+    }
+
+    fn finish_progress(&self, progress_bar_id: usize) {
+      let mut internal_state = self.state.write();
+
+      if let Some(index) = internal_state
+        .progress_bars
+        .iter()
+        .position(|p| p.id == progress_bar_id)
+      {
+        internal_state.progress_bars.remove(index);
+      }
+
+      // Remove last printed line
+      clear_previous_line();
+    }
+
+    fn start_draw_thread(&self, internal_state: &mut InternalState) {
+      internal_state.drawer_id += 1;
+      let drawer_id = internal_state.drawer_id;
+      let internal_state = self.state.clone();
+      tokio::task::spawn_blocking(move || {
+        loop {
+          {
+            let internal_state = internal_state.read();
+            // exit if not the current draw thread or there are no more progress bars
+            if internal_state.drawer_id != drawer_id
+              || internal_state.progress_bars.is_empty()
+            {
+              break;
+            }
+
+            let terminal_width = get_terminal_width().unwrap();
+            let mut text = String::new();
+            let progress_bar = &internal_state.progress_bars[0];
+
+            text.push_str(&get_progress_bar_text(
+              terminal_width,
+              *progress_bar.pos.read(),
+              progress_bar.size,
+              progress_bar.style,
+              progress_bar.start_time.elapsed().unwrap(),
+            ));
+
+            eprint!("{}", text);
+          }
+
+          std::thread::sleep(Duration::from_millis(100));
```

Comment:
Make it 16 for that sweet 60fps?

Reply:
Slow computers might struggle. I think we might have to add some more code that times how long a draw takes then update the fps based on that if we want to go down to 16.